### PR TITLE
test: Use both of gcc and clang by default

### PR DESCRIPTION
**Pull request by ggstone0523**
Change default compiler to gcc and clang
so that if not using -c or --compiler option,
runtest.py use gcc and clang for default compiler
and display test result of using gcc and clang compiler.

Signed-off-by: Yusun Choi <tanks2438@outlook.kr>

### CODE REVIEW

File: tests/runtest.py [hunk 1]
```diff
@@ -808,8 +814,8 @@ def parse_argument():
                         help="fail test if it runs more than TIMEOUT seconds")
     parser.add_argument("-j", "--worker", dest='worker', type=int, default=multiprocessing.cpu_count(),
                         help="Parallel worker count; using all core for default")
-    parser.add_argument("-c", "--compiler", dest='compiler', default="gcc",
-                        help="Select compiler gcc or clang. (gcc by default)")
+    parser.add_argument("-c", "--compiler", dest='compiler', default="gccclang",
```
Comment (honggyukim) [inline tests/runtest.py:817]: I think `all` is better than `gccclang`.

File: tests/runtest.py [hunk 2]
```diff
@@ -845,6 +851,11 @@ def parse_argument():
     else:
         flags = arg.flags.split()
 
+    if arg.compiler == 'gccclang':
+        compilers = ['gcc', 'clang']
```
Comment (honggyukim) [inline tests/runtest.py:855]: We better check if `gcc` and `clang` are installed in the system. If not, we should use only an available compiler.

**Review by honggyukim — COMMENTED** [review]
Thanks for your work. At first glance, I would like to ask you to change the output.

Current:
```
$ ./runtest.py 001
Start 1 tests with 1 worker
Compiler                               GCC                          CLANG
Test case                 pg             finstrument-fu pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os O0 O1 O2 O3 Os O0 O1 O2 O3 Os
001 basic               : OK OK OK OK OK OK OK OK OK OK OK OK OK OK OK OK OK OK OK OK
```
After changes:
```
$ ./runtest.py 001
Start 1 tests with 1 worker
Compiler                  gcc                            clang
Test case                 pg             finstrument-fu  pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os  O0 O1 O2 O3 Os O0 O1 O2 O3 Os
001 basic               : OK OK OK OK OK OK OK OK OK OK  OK OK OK OK OK OK OK OK OK OK
```

**Comment by ggstone0523** [comment]
Thanks you for review! I amend commit message following the comment.

File: tests/runtest.py [hunk 3]
```diff
@@ -845,6 +850,18 @@ def parse_argument():
...
+        gcc_installed = os.system('command -v gcc > /dev/null') == 0
+        clang_installed = os.system('command -v clang > /dev/null') == 0
+        if gcc_installed and not clang_installed:
+            compilers = ['gcc']
+        elif clang_installed and not gcc_installed:
+            compilers = ['clang']
+        else:
+            compilers = ['gcc', 'clang']
+    else:
+        compilers = [arg.compiler]
```
[the remark this answers is not in the document]
Comment (ggstone0523) [reply tests/runtest.py]: I think it's good! so I amend commit following the comment.

**Comment by ggstone0523** [comment]
I apologize for not checking the reviews properly,
and I amend commit for add gap between gcc test result and clang test result

File: tests/runtest.py [hunk 4]
```diff
@@ -845,6 +861,22 @@ def parse_argument():
...
+    compilerslen = 0
+    if arg.compiler == 'all':
+        for compiler in ['gcc', 'clang']:
+            if has_compiler(compiler):
+                compilers.append(compiler)
+                compilerslen += 1
+    else:
+        if has_compiler(arg.compiler):
+            compilers.append(arg.compiler)
+            compilerslen += 1
```
Comment (namhyung) [inline tests/runtest.py:878]: Do we really need to calculate the length separately?  Why not using `len(compilers)`?

File: tests/runtest.py [hunk 5]
```diff
@@ -808,8 +824,8 @@ def parse_argument():
                         help="fail test if it runs more than TIMEOUT seconds")
     parser.add_argument("-j", "--worker", dest='worker', type=int, default=multiprocessing.cpu_count(),
                         help="Parallel worker count; using all core for default")
-    parser.add_argument("-c", "--compiler", dest='compiler', default="gcc",
-                        help="Select compiler gcc or clang. (gcc by default)")
+    parser.add_argument("-c", "--compiler", dest='compiler', default="all",
+                        help="Select compiler gcc or clang. (gcc and clang by default)")
```
Comment (namhyung) [inline tests/runtest.py:828]: I'd say "(use both by default)".

**Review by namhyung — COMMENTED** [review]
I think the commit message is a little bit confusing.  It should say that it'd use all the available compilers.  How about "Use both of gcc and clang by default"?

**Comment by ggstone0523** [comment]
Thanks you for review! I amend commit following comment.

**Comment by honggyukim** [comment]
Could you take the following change and update this PR?
```diff
diff --git a/tests/runtest.py b/tests/runtest.py
index 48d5ca72..241d11a3 100755
--- a/tests/runtest.py
+++ b/tests/runtest.py
@@ -721,7 +721,7 @@ def save_test_result(result, case, shared):
         shared.total += 1


-def print_test_result(case, result, diffs, color, ftests, compilerslen):
+def print_test_result(case, result, diffs, color, ftests, nr_compilers):
     plain_result = [text_result[r] for r in result]

     if color:
@@ -735,12 +735,14 @@ def print_test_result(case, result, diffs, color, ftests, compilerslen):

     output = case[1:4]
     output += ' %-20s' % case[5:] + ':'
-    result_per_compiler = len(result_list) // compilerslen
-    for i in range(compilerslen):
+    result_per_compiler = len(result_list) // nr_compilers
+    for i in range(nr_compilers):
         output += ' '
         if i != 0:
             output += ' '
-        output += ' '.join(result_list[result_per_compiler*i:result_per_compiler*(i+1)])
+        begin = result_per_compiler * i
+        end = result_per_compiler * (i + 1)
+        output += ' '.join(result_list[begin:end])
     output += '\n'

     sys.stdout.write(output)
@@ -873,7 +875,10 @@ if __name__ == "__main__":
     else:
         if has_compiler(arg.compiler):
             compilers.append(arg.compiler)
-    compilerslen = len(compilers)
+    nr_compilers = len(compilers)
+    if nr_compilers == 0:
+        print('no compilers available for testing\n')
+        sys.exit(-1)

     from functools import partial
     class dotdict(dict):
@@ -936,7 +941,7 @@ if __name__ == "__main__":
             save_test_result(results, case=name, shared=shared)

             # Print sequentially when executing in serial
-            print_test_result(name, shared.results[name], shared.diffs[name], color, ftests, compilerslen)
+            print_test_result(name, shared.results[name], shared.diffs[name], color, ftests, nr_compilers)

     if use_pool:
         # Print via polling when using multiprocessing pool
@@ -946,7 +951,7 @@ if __name__ == "__main__":
             while name not in shared.results:
                 time.sleep(1)

-            print_test_result(name, shared.results[name], shared.diffs[name], color, ftests, compilerslen)
+            print_test_result(name, shared.results[name], shared.diffs[name], color, ftests, nr_compilers)

     if use_pool:
         pool.close()
```

**Comment by ggstone0523** [comment]
Thanks you for comment! I amend commit following comment.

**Review by honggyukim — APPROVED** [review]
Thanks for the update. LGTM.

**Pull request merged**